File: 743. Network Delay Time.cpp
using namespace std;

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        unordered_map<int, vector<pair<int, int>>> adj;
        vector<bool> visited(n + 1, false);

        for (const auto& time : times) {
            adj[time[0]].emplace_back(time[1], time[2]);
        }

        // Min-heap priority queue: pair<vertex, cost>, ordered by cost
        priority_queue<pair<int, int>, vector<pair<int, int>>, 
                       function<bool(pair<int,int>, pair<int,int>)>> q(
            [](const pair<int,int>& a, const pair<int,int>& b) {
                return a.second > b.second;
            }
        );

        q.emplace(k, 0);

        int ans = 0;
        while (!q.empty()) {
            auto [vertex, cost] = q.top();
            q.pop();

            if (visited[vertex]) continue;
            visited[vertex] = true;
            ans = max(ans, cost);

            if (adj.find(vertex) == adj.end()) continue;

            for (const auto& neig : adj[vertex]) {
                if (!visited[neig.first]) {
                    q.emplace(neig.first, cost + neig.second);
                }
            }
        }

        for (int i = 1; i <= n; i++) {
            if (!visited[i]) return -1;
        }

        return ans;
    }
};

File: 743. Network Delay Time.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[1], b[1]));
        HashMap<Integer, ArrayList<int[]>> adj = new HashMap<>();
        for(int[] time : times){
            ArrayList<int[]> temp = adj.getOrDefault(time[0], new ArrayList<>());
            temp.add(new int[]{time[1], time[2]});
            adj.put(time[0], temp);
        }

        pq.offer(new int[]{k, 0});
        int[] visited = new int[n + 1];
        Arrays.fill(visited, -1);
        visited[k] = 0;
        while(!pq.isEmpty()){
            int size = pq.size();
            for(int i = 0; i < size; i++){
                int node = pq.peek()[0];
                int cost = pq.poll()[1];

                if(!adj.containsKey(node)) continue;
                for(int[] neig : adj.get(node)){
                    int nextNode = neig[0];
                    int nextCost = neig[1] + cost;

                    if(visited[nextNode] == -1 || visited[nextNode] > nextCost){
                        visited[nextNode] = nextCost;
                        pq.offer(new int[]{nextNode, nextCost});
                    }
                }
            }
        }

        int ans = 0;
        for(int i = 1; i <= n; i++){
            if(visited[i] == -1) return -1;
            ans = Math.max(ans, visited[i]);
        }

        return ans;
    }
}